File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/local/StarshipDao.java
package com.dani_chuks.andeladeveloper.starwars.data.db.local;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Starship;
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.Vehicle;

import java.util.List;

import io.reactivex.Flowable;

@Dao
public interface StarshipDao {
    @Query("SELECT * FROM starship")
    Flowable<List<Starship>> getAll();

    @Query("SELECT * FROM starship LIMIT 7")
    Flowable<List<Starship>> getFirstSeven();

    @Query("SELECT * FROM starship ORDER BY name ASC" )
    Flowable<List<Starship>> getAllAlphabetically();

    @Query("SELECT * FROM starship WHERE name = :starshipName")
    Flowable <Starship> getPeopleByName(String starshipName);

    @Query("SELECT * FROM starship WHERE url = :starshipUrl")
    Flowable<Starship> getStarshipByURL(String starshipUrl);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertStarshipList(List<Starship> starships);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertStarship(Starship starship);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updateS(Vehicle vehicle);
}

File: app/src/main/java/com/dani_chuks/andeladeveloper/starwars/data/db/local/dao/StarshipDao.kt
package com.dani_chuks.andeladeveloper.starwars.data.db.local.dao

import androidx.room.Dao
import androidx.room.Query
import com.dani_chuks.andeladeveloper.starwars.data.models.entities.StarShip
import kotlinx.coroutines.flow.Flow

@Dao
abstract class StarshipDao: BaseDao<StarShip> {
    @Query("SELECT * FROM starship")
    abstract suspend fun all(): List<StarShip>

    @Query("SELECT * FROM starship ORDER BY name ASC")
    abstract suspend fun allAlphabetically(): List<StarShip>

    @Query("SELECT * FROM starship LIMIT :size")
    abstract suspend fun getItemBySize(size: Int): List<StarShip>

    @Query("SELECT * FROM starship WHERE name = :starshipName")
    abstract suspend fun getPeopleByName(starshipName: String): StarShip

    @Query("SELECT * FROM starship WHERE url = :starshipUrl")
    abstract suspend fun getStarshipByURL(starshipUrl: String): StarShip

    @Query("SELECT * FROM starship")
    abstract fun allAsFlow(): Flow<List<StarShip>?>

    @Query("SELECT * FROM starship ORDER BY name ASC")
    abstract fun allAlphabeticallyAsFlow(): Flow<List<StarShip>?>

    @Query("SELECT * FROM starship LIMIT :size")
    abstract fun getStarShipsBySizeAsFlow(size: Int): Flow<List<StarShip>?>

    @Query("SELECT * FROM starship WHERE name = :starshipName")
    abstract fun getStarshipByNameAsFlow(starshipName: String): Flow<StarShip?>

    @Query("SELECT * FROM starship WHERE url = :starshipUrl")
    abstract fun getStarshipByURLAsFlow(starshipUrl: String): Flow<StarShip?>
}
